src/main/java/net/dries007/tfc/common/blockentities/IFarmland.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed under the EUPL, Version 1.2.
+ * You may obtain a copy of the Licence at:
+ * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
+ */
+
+package net.dries007.tfc.common.blockentities;
+
+import java.util.List;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.chat.Component;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.util.Mth;
+
+import net.dries007.tfc.client.particle.TFCParticles;
+import net.dries007.tfc.util.Fertilizer;
+import net.dries007.tfc.util.Helpers;
+import net.dries007.tfc.common.blockentities.FarmlandBlockEntity.NutrientType;
+import static net.dries007.tfc.common.blockentities.FarmlandBlockEntity.NutrientType.*;
+
+/**
+ * Implement on block entities that hold nitrogen, phosphorous, and potassium values to allow fertilization and consuming nutrients.
+ */
+public interface IFarmland
+{
+    static void addNutrientParticles(ServerLevel level, BlockPos pos, Fertilizer fertilizer)
+    {
+        final float n = fertilizer.getNitrogen(), p = fertilizer.getPhosphorus(), k = fertilizer.getPotassium();
+        for (int i = 0; i < (int) (n > 0 ? Mth.clamp(n * 10, 1, 5) : 0); i++)
+        {
+            level.sendParticles(TFCParticles.NITROGEN.get(), pos.getX() + level.random.nextFloat(), pos.getY() + level.random.nextFloat() / 5D, pos.getZ() + level.random.nextFloat(), 0, 0D, 0D, 0D, 1D);
+        }
+        for (int i = 0; i < (int) (p > 0 ? Mth.clamp(p * 10, 1, 5) : 0); i++)
+        {
+            level.sendParticles(TFCParticles.PHOSPHORUS.get(), pos.getX() + level.random.nextFloat(), pos.getY() + level.random.nextFloat() / 5D, pos.getZ() + level.random.nextFloat(), 0, 0D, 0D, 0D, 1D);
+        }
+        for (int i = 0; i < (int) (k > 0 ? Mth.clamp(k * 10, 1, 5) : 0); i++)
+        {
+            level.sendParticles(TFCParticles.POTASSIUM.get(), pos.getX() + level.random.nextFloat(), pos.getY() + level.random.nextFloat() / 5D, pos.getZ() + level.random.nextFloat(), 0, 0D, 0D, 0D, 1D);
+        }
+    }
+
+    float getNutrient(NutrientType type);
+
+    void setNutrient(NutrientType type, float value);
+
+    default void addNutrient(NutrientType type, float value)
+    {
+        setNutrient(type, getNutrient(type) + value);
+    }
+
+    default void addNutrients(Fertilizer fertilizer)
+    {
+        setNutrient(NITROGEN, Math.min(1, getNutrient(NITROGEN) + fertilizer.getNitrogen()));
+        setNutrient(PHOSPHOROUS, Math.min(1, getNutrient(PHOSPHOROUS) + fertilizer.getNitrogen()));
+        setNutrient(POTASSIUM, Math.min(1, getNutrient(POTASSIUM) + fertilizer.getNitrogen()));
```

Comment:
`min` call is not necessary here as it's calling `setNutrient` which should be clamping the value anyway.